File: m26-py/m26/age.py
__author__ = 'cjoakim'


class Age(object):
    """
    Instances of this class represent the age of a person, as a float value
    of their age in years.
    """
    def __init__(self, n=0.0):
        self.value = float(n)

    def max_pulse(self):
        if self.value < 20:
            return 200.0
        else:
            return 220.0 - self.value

    def add(self, another_instance):
        if another_instance:
            self.value = self.value + another_instance.value
        return self.value

    def subtract(self, another_instance):
        if another_instance:
            self.value = self.value - another_instance.value
        return self.value

    def training_zones(self):
        results = list()
        zones = [0.95, 0.90, 0.85, 0.80, 0.75]
        for idx, pct in enumerate(zones):
            data = dict()
            data['zone'] = idx + 1
            data['age'] = self.value
            data['max'] = self.max_pulse()
            data['pct_max'] = pct
            data['pulse'] = self.max_pulse() * pct
            results.append(data)
        return results

    def __str__(self):
        return "<Age value:{0}>".format(self.value)

File: m26-js-ts/src/Age.ts

import { TrainingZone } from "./TrainingZone";

export class Age {
    
    constructor(public value: number) {

    }

    maxPulse(): number {
        if (this.value <= 20.0) {
            return 200;
        }
        else {
            return 220 - this.value
        }
    }

    add(another: Age): number {
        this.value = this.value + another.value;
        return this.value;
    }

    subtract(another: Age): number {
        this.value = this.value - another.value;
        if (this.value < 0) {
            this.value = 0;
        }
        return this.value;
    }

    trainingZones() : Array<TrainingZone> {
        let pctsOfMax = [ 0.75, 0.80, 0.85, 0.90, 0.95 ];
        let zones = Array();
        let max = this.maxPulse();
        for (let i = 0; i < pctsOfMax.length; i++) {
            let pctOfMax = pctsOfMax[i];
            zones.push(new TrainingZone(
                i+1, pctOfMax, this.value, max, Math.round(max * pctOfMax)));
        }
        return zones;
    }
}
